### PR TITLE
Set initial stack pointer using stackRestore

### PR DESCRIPTION
**Pull request by sbc100**
Don't assume its stored in linear memory.  When using
wasm32-unknown-unkown-wasm we use a wasm global to model
the stack pointer.

This code should work for both s2wasm and lld use cases.

### CODE REVIEW

File: src/preamble.js [hunk 1]
```diff
@@ -2187,6 +2187,7 @@ function integrateWasmJS() {
 #else
       removeRunDependency('wasm-instantiate');
 #endif
+      stackRestore(STACKTOP);
```
Comment (kripken) [inline src/preamble.js]: lgtm but please add a comment here, something like `set up the initial position of the stack`.

Comment (kripken) [reply src/preamble.js]: Also maybe add `TODO: it is not necessary to pass in STACKTOP as an import when this is used`. Which is in the wasm backend case, so maybe this should be in `#ifdef WASM_BACKEND`.

Comment (jgravelle-google) [reply src/preamble.js]: Seconding `#if WASM_BACKEND`

**Comment by sbc100** [comment]
Updated, and tested this time.  PTAL

File: src/preamble.js [hunk 2]
```diff
@@ -2179,6 +2179,9 @@ function integrateWasmJS() {
       if (exports.memory) mergeMemory(exports.memory);
       Module['asm'] = exports;
       Module["usingWasm"] = true;
+#if WASM_BACKEND
+      Module["_stackRestore"](STACKTOP);
```
Comment (kripken) [inline src/preamble.js]: Hmm, this doesn't look right to me. First, I think it should be exported without the `_` prefix. Also, it isn't necessarily exported (we no longer export all the runtime by default). Instead, calling `stackRestore` directly (not on Module) should work, like it was before - why did you change it from that? I may be missing something here.

Comment (sbc100) [reply src/preamble.js]: Hmm.. I changed it because I was getting a few test failure that looked like this:

```
+TypeError: stackRestore is not a function
+    at receiveInstance (/tmp/tmpyg29kY/emscripten_test_binaryen2_aNCc2n/src.cpp.o.js:1:17965)
+    at doNativeWasm (/tmp/tmpyg29kY/emscripten_test_binaryen2_aNCc2n/src.cpp.o.js:1:18777)
+    at Object.Module.asm (/tmp/tmpyg29kY/emscripten_test_binaryen2_aNCc2n/src.cpp.o.js:1:20161)
+    at Object.<anonymous> (/tmp/tmpyg29kY/emscripten_test_binaryen2_aNCc2n/src.cpp.o.js:1:106527)
+    at Module._compile (module.js:573:30)
+    at Object.Module._extensions..js (module.js:584:10)
+    at Module.load (module.js:507:32)
+    at tryModuleLoad (module.js:470:12)
+    at Function.Module._load (module.js:462:3)
+    at Function.Module.runMain (module.js:609:10)
```

e.g. test_relocatable_void_function

Comment (sbc100) [reply src/preamble.js]: Turns out the other approach also gives test failures too. so I don't have a working solution yet.

Comment (kripken) [reply src/preamble.js]: Hmm, in emscripten.py I see
```
  # some runtime functionality may not have been generated in
  # the wasm; provide a JS shim for it
  for name in ['setTempRet0', 'getTempRet0', 'stackSave', 'stackRestore', 'stackAlloc']:
    if name not in exported_implemented_functions:
      module.append('var %s;\n' % name)
```
The last part generates a `var stackRestore` is the wasm doesn't contain it (necessary for closure compiler). Which suggests that the wasm backend path doesn't always generate it - not sure why that would be?

Comment (sbc100) [reply src/preamble.js]: With -O0:

```
/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:110
+      throw ex;
+      ^
+abort("cannot use the stack before compiled code is ready to run, and has provided stack access") at Error
+    at jsStackTrace (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1190:13)
+    at stackTrace (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1207:12)
+    at abort (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:2811:44)
+    at getTempRet0 (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:226:3)
+    at receiveInstance (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1821:7)
+    at doNativeWasm (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1848:5)
+    at Object.Module.asm (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:1927:15)
+    at Object.<anonymous> (/tmp/emscripten_test_binaryen0_NqSohl/src.cpp.o.js:2237:24)
+    at Module._compile (module.js:573:30)
+    at Object.Module._extensions..js (module.js:584:10)

```

Comment (sbc100) [reply src/preamble.js]: Seems like it fails when BINARYEN_ASYNC_COMPILATION is off and some tests force this to be off..

**Comment by sbc100** [comment]
Ok, finally fixed all the cases I could find and documented why I can't use stackRestore() directly.

File: src/preamble.js (hunk 2 above)
Comment (kripken) [reply src/preamble.js]: Ah good, that `cannot use the stack before compiled code is ready to run` assertion in O0 may explain what's going on here. We are still in the process of setting things up for compiled code, and calling into it here may be dangerous. In practice it isn't here, so we can either add an exception for this function, or we can do this call in a later time when the runtime is ready. I think that's best, with something like this:
```
addOnInit(function() {
  stackRestore(STACKTOP);
});
```

File: src/preamble.js [hunk 3]
```diff
@@ -2179,6 +2179,14 @@ function integrateWasmJS() {
       if (exports.memory) mergeMemory(exports.memory);
       Module['asm'] = exports;
       Module["usingWasm"] = true;
+#if WASM_BACKEND
+      // wasm backend stack goes down
+      STACKTOP = STACK_BASE + TOTAL_STACK;
+      STACK_MAX = STACK_BASE;
+      // can't call stackRestore() here since this function can be called
+      // synchronously before stackRestore() is declared.
+      Module["asm"]["stackRestore"](STACKTOP);
```
Comment (kripken) [inline src/preamble.js]: or, ignore my last comment, doing it here may be nicer, but how about
```
  // We can now set the stack's initial position in the wasm module.
  exports['stackRestore'](STACKTOP);
```
Using `exports` directly clarifies what's going on, I think - we call directly into the exports that we just received.

Comment (sbc100) [reply src/preamble.js]: I think i had trouble with the metaDCE discarding the wam stackRestore function in that case.   Or at least with one of the ways I called it.   Will metaDCE see `exports['stackRestore'](STACKTOP);`?

Comment (kripken) [reply src/preamble.js]: Ah, good point. Yes, it's unsafe to capture the exports and access them like I suggested, metadce won't see it.

lgtm!
